docs/development/design/build-images.rst

```diff
@@ -48,131 +45,242 @@ This produced issues to people pinning their images to any of these names becaus
 * new images won't contain old/deprecated OS (eg. Ubuntu 18) and Python versions (eg. 3.5, miniconda2)
+* install language requirements *at built time* using ``asdf`` and its plugins
+* pre-build images with most common languages combination used by users to speed up built times
+* create local mirrors for all languages supported
+* deleting a pre-built image won't make builds to fail; only make them slower
+* support only the latest Ubuntu LTS version and give 2 year to users to migrate from the previous one
+* reduce the number of images supported only to the latest 2 Ubuntu LTS versions
 
 
 Non goals
 ---------
 
 * allow creation/usage of custom Docker images
 * allow to execute arbitrary commands via hooks (eg. ``pre_build``)
+* automatically build & push *all* images on commit
+* pre-built multiple images for all the languages combinations
 
 
-New build image structure
--------------------------
+Pre-built build image structure
+-------------------------------
 
 .. Taken from https://github.com/readthedocs/readthedocs-docker-images/blob/master/Dockerfile
 
+The new pre-built images will depend only on the Ubuntu OS.
+They will contain all the requirements to add extra languages support at built time via ``asdf`` command.
+
 * ``ubuntu20-base``
 
   * labels
   * environment variables
   * system dependencies
   * install requirements
   * LaTeX dependencies (for PDF generation)
-  * other languages version managers (``pyenv``, ``nodenv``, etc)
+  * languages version manager (``asdf``) and its plugins for each language
   * UID and GID
 
-The following images all are based on ``ubuntu20-base``:
+Instead of building all the Docker image per language versions combination,
+it will be easier to install all of them at build time using the same steps.
+Installing a language only adds a few seconds when binaries are provided.
+However, to reduce the time to install these languages as much as possible,
+a local mirror hosted on S3 for each language will be created.
 
-* ``ubuntu20-py*``
+It's important to note that Python does not provide binaries and compiling a version takes around ~2 minutes.
+However, the Python versions could be pre-compiled and expose their binaries via S3 to builders.
+Then, at build time, the builder will only download the binary and copy it in the correct path.
 
-  * Python version installed via ``pyenv``
-  * default Python packages (pinned versions)
-    * pip
-    * setuptools
-    * virtualenv
-  * labels
-
-* ``ubuntu20-conda*``
+.. note::
 
-  * same as ``-py*`` versions
-  * Conda version installed via ``pyenv``
-  * ``mamba`` executable (installed via ``conda``)
+   Depending on the demand, Read the Docs may pre-build the most common combinations of languages used by users.
+   For example, ``ubuntu20-python39+node14`` or ``ubuntu20-python39+node14+rust1``.
+   However, this is seen as an optimization for the future and it's not required for this document.
```

Comment:
Sounds like we might be in agreement on the plan forward, but some of the language in this doc still sounds like the plan is to build these language versions into the docker images. Just to clarify, are docker images with pre-built language versions included part of the plan in this doc still?

Reply:
> are docker images with pre-built language versions included part of the plan in this doc still?

No.

However, as we talked too much about this and we considered a "good idea" in different opportunities, I left this idea opened to be implemented in the future in case we decide to.

I'm not thinking on building Docker images with pre-installed languages on it, at the moment. I think the best decision is to move forward with the mirror cache and pre-compiled Python versions cache that we will install at build time.